File: solutions/java/src/main/java/com/epi/LargestRectangleUnderSkyline.java
package com.epi;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;

public class LargestRectangleUnderSkyline {
  public static int calculateLargestRectangleAlternative(List<Integer> A) {
    // Calculate L.
    Deque<Integer> s = new LinkedList<>();
    List<Integer> L = new ArrayList<>();
    for (int i = 0; i < A.size(); ++i) {
      while (!s.isEmpty() && A.get(s.peekFirst()) >= A.get(i)) {
        s.removeFirst();
      }
      L.add(s.isEmpty() ? Integer.valueOf(-1) : s.peekFirst());
      s.addFirst(i);
    }

    // Clear stack for calculating R.
    while (!s.isEmpty()) {
      s.removeFirst();
    }
    List<Integer> R = new ArrayList<>(Collections.nCopies(A.size(), 0));
    for (int i = A.size() - 1; i >= 0; --i) {
      while (!s.isEmpty() && A.get(s.peekFirst()) >= A.get(i)) {
        s.removeFirst();
      }
      R.set(i, s.isEmpty() ? A.size() : s.peekFirst());
      s.addFirst(i);
    }

    // For each A.get(i), find its maximum area include it.
    int maxArea = 0;
    for (int i = 0; i < A.size(); ++i) {
      maxArea = Math.max(maxArea, A.get(i) * (R.get(i) - L.get(i) - 1));
    }
    return maxArea;
  }

  // @include
  public static int calculateLargestRectangle(List<Integer> A) {
    Deque<Integer> s = new LinkedList<>();
    int maxArea = 0;
    for (int i = 0; i <= A.size(); ++i) {
      while (!s.isEmpty()
             && (i == A.size() || A.get(i) < A.get(s.peekFirst()))) {
        int height = A.get(s.peekFirst());
        s.removeFirst();
        maxArea = Math.max(maxArea,
                           height * (s.isEmpty() ? i : i - s.peekFirst() - 1));
      }
      s.addFirst(i);
    }
    return maxArea;
  }
  // @exclude

  // O(n^2) implementation checks answer.
  private static int checkAnswer(List<Integer> A) {
    int max = -1;
    for (int i = 0; i < A.size(); ++i) {
      int left = i - 1, right = i + 1;
      while (left >= 0 && A.get(left) >= A.get(i)) {
        --left;
      }
      while (right < A.size() && A.get(right) >= A.get(i)) {
        ++right;
      }
      int area = (right - left - 1) * A.get(i);
      if (area > max) {
        max = area;
      }
    }
    System.out.println(max);
    return max;
  }

  private static void smallTest() {
    List<Integer> A = Arrays.asList(2, 3, 4, 1, 2);
    int area = calculateLargestRectangle(A);
    int alterArea = calculateLargestRectangleAlternative(A);
    assert(area == alterArea);
    assert(checkAnswer(A) == area);
    assert(6 == area);
    A = Arrays.asList(2, 2, 2);
    System.out.println(calculateLargestRectangle(A));
    assert(6 == calculateLargestRectangle(A));
  }

  public static void main(String[] args) {
    smallTest();
    Random r = new Random();
    for (int times = 0; times < 3000; ++times) {
      int n;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = r.nextInt(1000) + 1;
      }
      List<Integer> A = new ArrayList<>();
      for (int i = 0; i < n; ++i) {
        A.add(r.nextInt(999));
      }
      int area = calculateLargestRectangle(A);
      int alterArea = calculateLargestRectangleAlternative(A);
      System.out.println(area + " " + alterArea);
      assert(area == alterArea);
      assert(checkAnswer(A) == area);
    }
  }
}

File: solutions/cpp/Largest_rectangle_under_skyline.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <random>
#include <vector>

#include "./Largest_rectangle_under_skyline.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// O(n^2) implementation checks answer.
int CheckAnswer(const vector<int>& A) {
  int max = -1;
  for (int i = 0; i < A.size(); ++i) {
    int left = i - 1, right = i + 1;
    while (left >= 0 && A[left] >= A[i]) {
      --left;
    }
    while (right < A.size() && A[right] >= A[i]) {
      ++right;
    }
    int area = (right - left - 1) * A[i];
    if (area > max) {
      max = area;
    }
  }
  cout << max << endl;
  return max;
}

void SmallTest() {
  vector<int> A = {2, 3, 4, 1, 2};
  int area = CalculateLargestRectangle(A);
  int alter_area = CalculateLargestRectangleAlternative(A);
  assert(area == alter_area);
  assert(CheckAnswer(A) == area);
  assert(6 == area);
  A = {2, 2, 2};
  assert(6 == CalculateLargestRectangle(A));
  A = {1, 1, 2};
  assert(3 == CalculateLargestRectangle(A));
}

int main(int argc, char* argv[]) {
  SmallTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 3000; ++times) {
    vector<int> A;
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 1000);
      n = dis(gen);
    }
    while (n--) {
      uniform_int_distribution<int> dis(0, 999);
      A.emplace_back(dis(gen));
    }
    int area = CalculateLargestRectangle(A);
    int alter_area = CalculateLargestRectangleAlternative(A);
    cout << area << " " << alter_area << endl;
    assert(area == alter_area);
    assert(CheckAnswer(A) == area);
  }
  return 0;
}
